server/src-lib/Hasura/RQL/DML/RemoteJoin.hs

```diff
@@ -0,0 +1,446 @@
+
+    -- | 'FieldCall' is path to remote relationship value in remote server response.
+    -- 'extractAtPath' traverse through the path and extracts the json value
+    extractAtPath path v =
+      case NE.nonEmpty path of
+        Nothing          -> pure v
+        Just (h :| rest) -> case v of
+          AO.Object o   -> maybe
+                           (throw500 $ "cannnot find value in remote response at path " <> T.pack (show path))
+                           (extractAtPath rest)
+                           (AO.lookup (G.unName h) o)
+          AO.Array arr -> AO.array <$> mapM (extractAtPath path) (toList arr)
+          _            -> throw500 $ "expecting array or object in remote response at path " <> T.pack (show path)
+
+-- | Fold nested 'FieldCall's into a bare 'Field', inserting the passed
+-- selection set at the leaf of the tree we construct.
+fieldCallsToField
+  :: MonadError QErr m
+  => [G.Argument]
+  -> Map.HashMap G.Variable G.Value
+  -> [G.Field]
+  -- ^ Inserted at leaf of nested FieldCalls
+  -> G.Alias
+  -- ^ Top-level name to set for this Field
+  -> NonEmpty FieldCall
+  -> m G.Field
+fieldCallsToField rrArguments variables finalSelSet topAlias =
+  fmap (\f -> f{G._fAlias = Just topAlias}) . nest
+  where
+    -- almost: `foldr nest finalSelSet`
+    nest ((FieldCall name remoteArgs) :| rest) = do
+      templatedArguments <- createArguments variables remoteArgs
+      (args, selSet) <- case NE.nonEmpty rest of
+            Just f -> do
+              s <- nest f
+              pure (templatedArguments, pure s)
+            Nothing ->
+              let argsToMap = Map.fromList . map (G._aName &&& G._aValue)
+                  arguments = map (uncurry G.Argument) $ Map.toList $
+                              Map.unionWith mergeValue
+                                (argsToMap rrArguments)
+                                (argsToMap templatedArguments)
+              in pure (arguments, finalSelSet)
+      pure $ G.Field Nothing name args [] $ map G.SelectionField selSet
+
+-- This is a kind of "deep merge".
+-- For e.g. suppose the input argument of the remote field is something like:
+-- `where: { id : 1}`
+-- And during execution, client also gives the input arg: `where: {name: "tiru"}`
+-- We need to merge the input argument to where: {id : 1, name: "tiru"}
+mergeValue :: G.Value -> G.Value -> G.Value
+mergeValue lVal rVal = case (lVal, rVal) of
+  (G.VList (G.ListValueG l), G.VList (G.ListValueG r)) ->
+    G.VList $ G.ListValueG $ l <> r
+  (G.VObject (G.ObjectValueG l), G.VObject (G.ObjectValueG r)) ->
+    let fieldsToMap = Map.fromList . map (G._ofName &&& G._ofValue)
+    in G.VObject $ G.ObjectValueG $ map (uncurry G.ObjectFieldG) $ Map.toList $
+       Map.unionWith mergeValue (fieldsToMap l) (fieldsToMap r)
+  (l, _) -> l -- FIXME:- throw error for merging non-lists and non-objects
+
+fieldsToRequest :: G.OperationType -> [G.Field] -> GQLReqParsed
```

Comment:
Since this is undocumented and used just once, can we move it into a where clause at the callsite in `fetchRemoteJoinFields`